### ISSUE TITLE
Experiment with skipping reconciliation for primitive components

### ISSUE THREAD

**Issue by LPGhatguy** [Roblox/roact#75]
Roblox deduplicates property changes already. What if we skipped the reconciliation step completely for primitives and set properties on every render?

This could be significantly faster in the worst cases. I wonder if it's ever slower?

**Comment by AmaranthineCodices** [comment]
I do wonder about that - it'd be worth a check. We can move the event deduplication to `SingleEventManager` if we want to keep that, but otherwise I think we can let Roblox handle deduplication without any issues.

**Comment by AmaranthineCodices** [comment]
After some benchmarks, this has **serious performance implications**: Disabling duplication actively harms performance in most cases I benchmarked, including the "realistic" scenario.

## Results
```
Benchmark reconcile_1.bench:
	(100000 iterations) took 4.004883 s (40048.832893 ns/iteration)
Benchmark reconcile_nodedup_1.bench:
	(100000 iterations) took 3.494400 s (34944.002628 ns/iteration)
Benchmark reconcile_2.bench:
	(100000 iterations) took 3.052541 s (30525.412560 ns/iteration)
Benchmark reconcile_nodedup_2.bench:
	(100000 iterations) took 3.289009 s (32890.090942 ns/iteration)
Benchmark reconcile_3.bench:
	(100000 iterations) took 6.186478 s (61864.776611 ns/iteration)
Benchmark reconcile_nodedup_3.bench:
	(100000 iterations) took 8.649346 s (86493.456364 ns/iteration)
```

## Benchmark steps

Here are the `step` blocks of the benchmarks. Note that `reconcile_#.bench` and `reconcile_#_nodedup.bench` have identical `step` blocks; the `_nodedup` benchmarks set a config flag to disable Roact's property deduplication for primitives.

### `reconcile_1`
`reconcile_1` and `reconcile_nodedup_1` benchmarked the case when the value changed every iteration. **This is the only case where disabling Roact's deduplication helped.** This makes sense, since we are doing unnecessary work in this case.
```lua
handle = Roact.reconcile(handle, Roact.createElement("StringValue", {
	Value = tostring(i),
}))
```

### `reconcile_2`
`reconcile_2` and `reconcile_nodedup_2` benchmarked the case where the value was static.
```lua
handle = Roact.reconcile(handle, Roact.createElement("StringValue", {
	Value = "Initial",
}))
```

### `reconcile_3`
`reconcile_3` and `reconcile_nodedup_3` benchmarked the more realistic case where only one property out of a set of many properties change.
```lua
handle = Roact.reconcile(handle, Roact.createElement("Frame", {
	Position = UDim2.new(0, 1, 0, i),
	BackgroundColor3 = Color3.new(1, 1, 1),
	BorderSizePixel = 3,
	BorderColor3 = Color3.new(1, 0, 0),
	AnchorPoint = Vector2.new(0.5, 0.5),
	Size = UDim2.new(0, 100, 0, 100),
	ZIndex = 4,
}))
```

## Deduplication-disabling code
```diff
+local dedup = GlobalConfig.getValue("primitiveDeduplication")
+
 -- Set properties that were set with fromElement
for key, oldValue in pairs(fromElement.props) do
	seenProps[key] = true

	local newValue = toElement.props[key]

	-- Assume any property that can be set to nil has a default value of nil
	if newValue == nil then
		local _, value = getDefaultPropertyValue(rbx.ClassName, key)

		-- We don't care if getDefaultPropertyValue fails, because
		-- _setRbxProp will catch the error below.
		newValue = value
	end

	-- Roblox does this check for normal values, but we have special
	-- properties like events that warrant this.
+	if not dedup or oldValue ~= newValue then
		Reconciler._setRbxProp(rbx, key, newValue, toElement)
	end
end

 -- Set properties that are new in toElement
for key, newValue in pairs(toElement.props) do
	if not seenProps[key] then
		seenProps[key] = true

		local oldValue = fromElement.props[key]

+		if not dedup or oldValue ~= newValue then
			Reconciler._setRbxProp(rbx, key, newValue, toElement)
		end
	end
end
```

## Conclusion

To summarize: This is detrimental to performance in most cases. My hypothesis is that the cost of going across the Lua bridge outweighs the benefits of letting Roblox deduplicate changes.

**Comment by LPGhatguy** [comment]
Darn. I think we can revisit this later after we've got an async reconciler.
